### ISSUE TITLE
[Question] Why does Embulk JRuby PageBuilder use `ValueFactory.newString(str)`, not `new JsonParser().parse(str)`, when parsing JSON string?

### ISSUE THREAD

**Issue by civitaspo** [embulk/embulk#499]
Hi maintainers,

I have a question that "Why does Embulk JRuby PageBuilder use `ValueFactory.newString(str)`, not `new JsonParser().parse(str)`, when parsing JSON string?".

When parsing JSON String in JRuby, our plugins just call [`PageBuilder#add`](https://github.com/embulk/embulk/blob/release_v0.8.14/lib/embulk/page_builder.rb#L16-L25) which finally calls `ValueFactory.newString(str)` in [`JsonColumnSetter#set`](https://github.com/embulk/embulk/blob/release_v0.8.14/embulk-core/src/main/java/org/embulk/spi/util/dynamic/JsonColumnSetter.java#L47-L51). On the other hand, in Java, most of all plugins call [`JsonParser#parse`](https://github.com/embulk/embulk/blob/release_v0.8.14/embulk-core/src/main/java/org/embulk/spi/json/JsonParser.java#L39-L42) then call [`PageBuilder#setJson`](https://github.com/embulk/embulk/blob/release_v0.8.14/embulk-core/src/main/java/org/embulk/spi/PageBuilder.java#L138-L151). So, when we create plugins then we parse a JSON String, the plugins would differ in subtle behaviour of between JRuby plugin and Java Plugin. JRuby plugins can set any `String` as "`JSON`" (moreover any data of Embulk data types can be set), though Java plugins call [`JsonParseException`](https://github.com/embulk/embulk/blob/release_v0.8.14/embulk-core/src/main/java/org/embulk/spi/json/JsonParseException.java) as long as the `String` is not `JSON`. That's why I have the question: "Why does Embulk JRuby PageBuilder use `ValueFactory.newString(str)`, not `new JsonParser().parse(str)`, when parsing JSON string?".

Needless to say, if the plugin authors use [`JsonParser#parse`](https://github.com/embulk/embulk/blob/release_v0.8.14/embulk-core/src/main/java/org/embulk/spi/json/JsonParser.java#L39-L42) in JRuby Plugins then use [`PageBuilder#add`](https://github.com/embulk/embulk/blob/release_v0.8.14/lib/embulk/page_builder.rb#L16-L25), the behaviour becomes the same, but I think using Java API directly is a "not less" load for JRuby plugin authors. So, in my opinion, Embulk should provide JRuby JsonParser API. How do you think?

(I thought an idea that [`JsonColumnSetter#set`](https://github.com/embulk/embulk/blob/release_v0.8.14/embulk-core/src/main/java/org/embulk/spi/util/dynamic/JsonColumnSetter.java#L47-L51) calls [`JsonParser#parse`](https://github.com/embulk/embulk/blob/release_v0.8.14/embulk-core/src/main/java/org/embulk/spi/json/JsonParser.java#L39-L42) in itself, but this way is too different from the way to use Java API and tightly coupled, so I did not think this is a good idea.)

**Comment by frsyuki** [comment]
I think it's OK to change the behavior. `JsonColumnSetter#set(String v)` can parse the given string as a JSON. 
But why don't you pass a Hash object to `Embulk::PageBuilder` so that `spi.util.dynamic.AbstractDynamicColumnSetter#setRubyObject(IRubyObject)` calls `JsonColumnSetter#set(Value)`?

**Comment by uu59** [comment]
I've always used `page_builder.add(hash)`, and never used `.add(json_string)`.
Because my typical usages are based on Web APIs. My plugins should parse a response as JSON at first. I have no chance to pass `add(json_string)` in my cases.

The typical API response like this:
```json
{
  "total_count": 123,
  "records": [...],
  "next_page_url": "https://..."
}
```

I can imagine use cases of `add(json_string)` (e.g. input from databases, files, etc).
But I have the same question of @frsyuki. `add(JSON.parse(json_string))` would be ok and deeply error handling can be done on the JRuby plugin side.

Thus, I disagree for "breaking" API changes. But it's okay to me if API accepts both of `add(hash)` and `add(json_string)`.

**Comment by civitaspo** [comment]
Thanks for the replies. Of course, I disagree with breaking API changes, too. Finally, I understand using `page_builder.add([JSON.parse(json_string)]` is the best way. So, I close this issue.

The reason why I said the question and the opinion `How does Embulk provide JsonParser Ruby API which Java implements?` is that the current Embulk JRuby implementation can input any values having `.to_msgpack` as JsonType so that, as Java JsonParser API, there should be the best way to implement about JsonType for not confusing. (Actually, I confused what value is acceptable, and I needed to read Java code for understanding it.) Anyway, I got the best way to implement and the intent of the implementation. Thanks so much!
